### ISSUE TITLE
Possibly too many commits displaying as uncommitted after a merge?

### ISSUE THREAD

**Issue by mikebeaton** [ForkIssues/TrackerWin#681]
I have this structure in my current project:

![image](https://user-images.githubusercontent.com/11946605/75185722-7e7e0b80-573e-11ea-88da-4586a633eeb0.png)

I would only expect blue dots at the merge commits on the orange branch, because these are the commits which have not been push to remote on `wip` or any other branch yet. I would not expect the other commits on the yellow `master` branch to show with a blue dot - as they are doing - because they *are* on the remote, on `master`.

I do realise the commits I'm talking about are not on `wip` on the remote yet - but still, it seems to me that they are on the remote, and I wouldn't expect them to show as blue. I would say that I find the current display slightly misleading, or less than perfect UI: looking at it, I was initially thinking, "How come there's some much uncommitted on `master`?"

I guess that tracking back, I would only expect blue dots for things that are not on the remote on any branch, and not for things that are not on the remote wrt to just some branches when they are on others.

**Comment by DanPristupov** [comment]
Those dots represent the ahead/behind info.

* Blue dots show commits ahead of the corresponding remote branch. I.e. commits which will appear in the remote branch after push.
* Gray dots show commits behind of the corresponding remote branch. I.e. commits which will appear in the local branch after pull.

**Comment by mikebeaton** [comment]
Yes, but the blue in this picture *aren't* ahead. Or rather, they are ahead on the orange branch, but *only* in virtue of having been merged in and the merges haven't been pushed. And they are *not* ahead on the yellow branch, whereas the UI makes it look very much as if they are. Do you see what I'm trying to say?

EDIT: to put it another way: master and origin/master **are** up-to-date, and yet there's loads of blue!

**Comment by DanPristupov** [comment]
Yes, I understand what you mean. However as I said, those dots are just the ahead/behind info between local branch and its upstream. They correspond to the info on the sidebar (it probably shows `wip 8↑` in your case). And they also correspond to the behind/ahead info on the toolbar.

> master and origin/master are up-to-date, and yet there's loads of blue!

The master branch is not related there. Only the active branch and the upstream are important.

**Comment by mikebeaton** [comment]
Ah, it's about the active branch.

Okay, this is probably not going to happen, but I guess I was mistakenly reading these dots as:

 - Blue - here's something we're highlighting for you, a commit that's in your local repo that you haven't put onto the remote yet
 - Grey - here's something else we'll show you, a commit that's in the remote but which hasn't been pulled into any of your local branches yet

To me that would arguably be even more useful to know?

But yes, I understand what you are saying about what it is doing currently now, yes it does show `wip 8↑`.

**Comment by mikebeaton** [comment]
Hi - I agree that this is not an issue as such; now that I've understood that the blue/grey-ness is determined with respect to the currently checked-out branch then I can see that what it's doing is completely correct.

But if you've got any time to answer this then, just to follow up the point above, do you think it's reasonable (e.g. not a stupid thing to want, with regard to how git fundamentally works!) to want to be able to see all the commits that aren't on _any_ remote branch yet, and all the remote commits that aren't on _any_ local branch yet? Is that something that could be made available a different way? Is it even something that I can already see in Fork, if I know what to look for?

Thanks!

**Comment by DanPristupov** [comment]
Yes this seems to be usable and I agree that in the majority of cases we only care whether commit is on remote server or not.

However it's not as simple as it sounds. Mostly because git is distributed and there's no default branch or remote.

1. A lot of people have more than one remote server. Whole pull request workflow on GitHub is based on having multiple remotes. When we have multiple remotes we can't say if branch is pushed (where?). We can only say if branch is up-to-date comparing to its remote tracking branch.
2. Some repos don't have remotes at all. We can't mark all commits with blue there.
3. This is how git works. We tend to not change the git behavior.

**Comment by mikebeaton** [comment]
Thank you for pointing out the potential problems here. What you say makes sense.

1. That's true, but you could still indicate whether a commit is out in public anywhere or not. i.e. If a commit is on multiple branches, then only show it as blue if it's not pushed to the relevant remote on any of them. Being on a branch with no remote counts as not pushed, I guess. In fact, I think what I've said is what I thought Fork was showing, which is why I (initially!) thought the display was wrong.
2. Definitely, if there were no remotes at all, then although logically (by what I'm asking for) all commits would all be blue, it would be preferable to just do none of this (no greys, no blues) until there is at least one remote. Of course, as soon as there is at least one remote, then if no local commits are up on it yet they would all go blue.
3. I'm definitely not asking to change how git works (and I can see this _may_ be skirting too close to that - I'm still not sure!).

I'm only following this up again because I _think_ it seems as if you're still at least slightly interested in whether this might be a) useful, b) do-able without going against how git works!

**Comment by mikebeaton** [comment]
I've always thought that git showing - e.g. - the `wip` branch in my screenshot as being 8 commits behind was kind of weird, since 'really' it's only the two merge commits behind.

I understand that that's probably fundamental to how git works, so that maybe I'm arguing against myself here(!), but to me it would be a really useful special feature of Fork if it DID only show the two merge commits as being 'blue' in my example, since (as I _think_ you almost agree!?) that is what we _really_ want to know. _Of course_ I'd still expect Fork to show `wip` as `8↑`, exactly as git reports it, even in this case - I'm definitely not suggesting changing that. ... I wonder if maybe there could be a hollow blue circle for 'behind on the current branch as git reports it' (everything showing blue in Fork now) and then use the filled blue circle for 'not up on its linked remote on _any_ branch' (only the two merge commits in my screenshot)??? That could possibly be very cool!

**Comment by mikebeaton** [comment]
Was there any chance of having different symbols (solid blue vs outlined blue) for "not on any remote branch" vs. "not on this remote branch"? :-)
